src/typescript-service.ts

```diff
@@ -417,6 +417,68 @@ export class TypeScriptService {
             })
     }
 
+    /**
+     * The goto type definition request is sent from the client to the server to resolve the type
+     * location of a symbol at a given text document position.
+     *
+     * @return Observable of JSON Patches that build a `Location[]` result
+     */
+    public textDocumentTypeDefinition(params: TextDocumentPositionParams, span = new Span()): Observable<Operation> {
+        return this._getTypeDefinitionLocations(params, span)
+            .map((location: Location): Operation => ({ op: 'add', path: '/-', value: location }))
+            .startWith({ op: 'add', path: '', value: [] })
+    }
+
+    /**
+     * Returns an Observable of the type definition location found for a symbol.
+     */
+    protected _getTypeDefinitionLocations(params: TextDocumentPositionParams, span = new Span()): Observable<Location> {
+        const uri = normalizeUri(params.textDocument.uri)
+
+        // Fetch files needed to resolve definition
+        return this.projectManager
+            .ensureReferencedFiles(uri, undefined, undefined, span)
+            .toArray()
+            .mergeMap(() => {
+                const fileName: string = uri2path(uri)
+                const configuration = this.projectManager.getConfiguration(fileName)
+                configuration.ensureBasicFiles(span)
+
+                const sourceFile = this._getSourceFile(configuration, fileName, span)
+                if (!sourceFile) {
+                    throw new Error(`Expected source file ${fileName} to exist`)
+                }
+
+                const offset: number = ts.getPositionOfLineAndCharacter(
+                    sourceFile,
+                    params.position.line,
+                    params.position.character
+                )
+                const definitions:
+                    | ts.DefinitionInfo[]
+                    | undefined = configuration.getService().getTypeDefinitionAtPosition(fileName, offset)
+
+                return Observable.from(definitions || []).map((definition): Location => {
+                    const sourceFile = this._getSourceFile(configuration, definition.fileName, span)
+                    if (!sourceFile) {
+                        throw new Error('expected source file "' + definition.fileName + '" to exist in configuration')
+                    }
+                    const start = ts.getLineAndCharacterOfPosition(sourceFile, definition.textSpan.start)
+                    const end = ts.getLineAndCharacterOfPosition(
+                        sourceFile,
+                        definition.textSpan.start + definition.textSpan.length
+                    )
+                    return {
+                        uri: locationUri(definition.fileName),
+                        range: {
+                            start,
+                            end,
+                        },
+                    }
+                })
+            })
+    }
```

Comment:
It seems like a lot of this implementation is copy-pasted except one line, would it maybe be possible to share to most the code?